### ISSUE TITLE
About specific position input 

### ISSUE THREAD

**Issue by docShen** [embodied-generalist/embodied-generalist#34]
Hello, thank you very much for your work. I am a doctoral student who has recently joined the research work of 3D agents. 

In this demo, I mentioned "Develop a description of the object at the specific position in the 3D backlog" and attached an object surrounded by a red bounding box in the 3D input. I would like to know how the "specific position" is understood by the network and how it is reflected in the code. I am confused. Looking forward to your reply!
![1721285992937](https://github.com/user-attachments/assets/f7fc1012-2283-47e5-9774-aabd0e385cd2)

**Comment by Germany321** [comment]
LEO does not predict the coordinates of the objects. The model directly provieds the text responses according to the 3D scene and the text prompt. In this demo, the bounding box is used for visualization. In fact, you can modify the model and add special tokens to predict coordinates of the objects.

**Comment by docShen** [comment]
> LEO does not predict the coordinates of the objects. The model directly provieds the text responses according to the 3D scene and the text prompt. In this demo, the bounding box is used for visualization. In fact, you can modify the model and add special tokens to predict coordinates of the objects.

Thank you for your reply. Sorry, I think I may not have described the problem clearly. 
What I mean is, in this example, how do I let LEO know that I am asking a question about a keyboard (not a chair, bed, etc.), and more specifically, how do I explicitly let LLM know that I am asking a token question about a keyboard (or any object).

My guess for this is to give LEO some additional location prompts, such as: "Please describe the keyboard and its neighbor object on the table in this scene." or: "Please describe what the object corresponding to the center position [x, y, z] is? And describe what is around it."  Is this how it is implemented? Or is there any other way? Thank you very much!

**Comment by Germany321** [comment]
> > LEO does not predict the coordinates of the objects. The model directly provieds the text responses according to the 3D scene and the text prompt. In this demo, the bounding box is used for visualization. In fact, you can modify the model and add special tokens to predict coordinates of the objects.
> 
> Thank you for your reply. Sorry, I think I may not have described the problem clearly. What I mean is, in this example, how do I let LEO know that I am asking a question about a keyboard (not a chair, bed, etc.), and more specifically, how do I explicitly let LLM know that I am asking a token question about a keyboard (or any object).
> 
> My guess for this is to give LEO some additional location prompts, such as: "Please describe the keyboard and its neighbor object on the table in this scene." or: "Please describe what the object corresponding to the center position [x, y, z] is? And describe what is around it." Is this how it is implemented? Or is there any other way? Thank you very much!

You can ask the question like "Please describe the keyboard and its neighbor object on the table in this scene.". But the question "Please describe what the object corresponding to the center position [x, y, z] is? And describe what is around it." may be difficult for LEO, since our curated dataset lacks such data, which will lead to the bias responses. Maybe you can add additional data in the training set to unlock this interesting capability.

**Comment by docShen** [comment]
I understand now. Thanks for your answer！

**Comment by jkstyle2** [comment]
@Germany321 @docShen 
I'm still confused. Do you mean that we need to prompt "Please describe the keyboard and its neighbor object on the table in this scene." given 3D input with the red bounding box? How can I reproduce the demo?

**Comment by huangjy-pku** [comment]
Let me clarify this. The *specific position* is associated with the model input whose key is `anchor_locs`, 3-dim coordinates sampling within the target object to indicate which object the model is supposed to describe. So, if you prompt in this manner such as *specific position*, you can modify the `anchor_locs` to customize what you want.

**Comment by jkstyle2** [comment]
@huangjy-pku 
I'm sorry to say that I'm still being confused.
Do you mean that the current published LEO is not able to take and predict any _specific position in 3D_ , but can be done when training with `anchor_locs`. Please correct me if I'm wrong.
However, in the paper below, the instruction asks LEO to describe about the object at 'specified position'. How is it possible?
![image](https://github.com/user-attachments/assets/0d094b0e-c6bf-4fff-bc3f-99a6e7d10fe7)

**Comment by huangjy-pku** [comment]
The key `anchor_locs` is not only used for training. It is always a part of the input. When the model is asked to describe *specific position*, it will utilize the information in `anchor_locs`, which indicates the coordinate of the *specific position*.

**Comment by jkstyle2** [comment]
I got it. So the `anchor_locs` is necessarily used for both training and predicting in captioning task for a given specific object.
Can I test it with `inference.sh` by passing a further argument `anchor_locs` to specify an object?
I also have a question why not using `obj_locs` or `obj_pcd` rather than `anchor_locs` to specifiy an object.
> anchor_loc = obj_pcd[:, :3].min(0) + np.random.rand(3) * obj_size

**Comment by huangjy-pku** [comment]
To generate captions for a specific object, you need to first obtain the location of that object and assign a corresponding position to `anchor_locs`.
